chainer/gradient_check.py

```diff
@@ -52,34 +85,182 @@ def numerical_grad(f, inputs, grad_outputs, eps=1e-3):
 
     if gpu:
         xp = cuda.cupy
-        numerical_grad_kernel = cuda.reduce(
+        numerical_grad_kernel_1 = cuda.reduce(
             'T y1, T y2, U gy, T eps', 'V gxi',
             '(y1 - y2) * gy', 'a + b', 'gxi += a / (eps * 2)', '0',
-            'numerical_grad_kernel'
+            'numerical_grad_kernel_1'
+        )
+        numerical_grad_kernel_3 = cuda.reduce(
+            'T y1, T y2, T y3, T y4, U gy, T eps', 'V gxi',
+            '(-y1 + 8 * y2 - 8 * y3 + y4) * gy',
+            'a + b', 'gxi += a / (eps * 6)', '0',
+            'numerical_grad_kernel_3'
         )
     else:
         xp = numpy
     grads = [xp.zeros(x.shape, numpy.float64) for x in inputs]
 
+    if detect_nondifferentiable:
+        if center_outputs is None:
+            ys0 = _copy_arrays(f())
+        else:
+            ys0 = center_outputs
+        nout = len(ys0)
+        shapes = [_.shape for _ in ys0]
+        sizes = numpy.array([_.size for _ in ys0])
+        cumsizes = numpy.cumsum(sizes)
+
+    # Evaluate func at a single input
+    def eval_func(x, i, delta, orig):
+        x[i] = orig + delta
+        y = _copy_arrays(f())
+        x[i] = orig
+        return y
+
+    # An iteration on a single input displacement
+    def iterate_single_input(i_in, x, orig_x, i):
+        orig = orig_x[i]
+        if detect_nondifferentiable:
+            yss = [
+                eval_func(x, i, -eps * 1., orig),
+                eval_func(x, i, -eps * .5, orig),
+                ys0,
+                eval_func(x, i, +eps * .5, orig),
+                eval_func(x, i, +eps * 1., orig),
+            ]
+        else:
+            yss = [
+                eval_func(x, i, -eps * 1, orig),
+                eval_func(x, i, +eps * 1, orig),
+            ]
+
+        if detect_nondifferentiable:
+            # Detect non-differentiable point by quadratic fitting
+
+            # Check for non-finite output.
+            # If any single element in the output arrays has different
+            # finiteness among sampled points, that means this is a
+            # non-differentiable point.
+            # If the function consistently generates non-finite values
+            # around the point, we do not treat the point as
+            # non-differentiable.
+            # (Example: x<0 region for the logarithm function)
+            any_nonfinite = False
+            for i_out in range(nout):
+                isfinites = [
+                    xp.isfinite(yss[_][i_out])
+                    for _ in range(len(yss))]
+                if any((isfinites[0] != isfinites[_]).any()
+                       for _ in range(1, len(yss))):
```

Comment:
Could you add a comment describing this condition?

Reply:
I've written that in the comment just above.
Isn't that clear?

Reply:
I thought that we could elaborate on the loop, even after reading the comment above. If that's too verbose, let's leave it as it is.

Reply:
I added some comments for `yss` and improves the code a bit. I hope it's easier to understand.

Reply:
Big thanks!